### ISSUE TITLE
Not receiving notifications and error on read() from Nordic UART Service peripheral

### ISSUE THREAD

**Issue by cahna** [innoveit/react-native-ble-manager#1034]
This issue is related to a peripheral (a "smart" stopwatch) that is documented as using [Nordic UART Service (NUS)](https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/libraries/bluetooth_services/services/nus.html) for bluetooth communication.

|Property|Value|
|---|---|
|Service UUID|6E400001-B5A3-F393-E0A9-E50E24DCCA9E|
|RX Characteristic|6E400002-B5A3-F393-E0A9-E50E24DCCA9E|
|TX Characteristic|6E400003-B5A3-F393-E0A9-E50E24DCCA9E|

The stopwatch documents that "commands" are sent via "simple human readable ascii text" and data/responses can be ASCII text or ASCII-encoded hex (for faster, parse-able responses).

**Describe the bug(s)**
1. Notifications for `BleEventType.BleManagerDidUpdateValueForCharacteristic` are never received (even after writing a "command" to RX Characteristic), __and/or__
2. After using `BleManager.write()` to write a "command", the subsequent `BleManager.read()` throws an error

**To Reproduce**
1. Android app Bluetooth setup/init/start:
    1. Check/request permissions, enable bluetooth, init&start `BleManager`
    2. Setup event listeners: `BleManagerStopScan`, `BleManagerDiscoverPeripheral`, `BleManagerConnectPeripheral`, `BleManagerDisconnectPeripheral`, `BleManagerDidUpdateState`, `BleManagerDidUpdateValueForCharacteristic`, & `BleManagerDidUpdateNotificationStateFor`
    3. Scan for serviceUUID: `6E400001-B5A3-F393-E0A9-E50E24DCCA9E`
        - [x] `BleManagerDiscoverPeripheral`  triggers when device is discovered
        - [x] `BleManagerStopScan` triggers at end of scan
    4. Call `BleManager.connect` with the discovered peripheral
        - [x] `BleManagerConnectPeripheral` triggers
    5. Retrieve peripheral services, as required before calling other methods. Returns (example):
        ```json
        {
          "id": "60:01:02:03:04:05",
          "name": "Stopwatch 1234",
          "rssi": -61,
          "advertising": {/* skipped */},
          "services": [
            { "uuid": "1800" },
            { "uuid": "1801" },
            { "uuid": "6e400001-b5a3-f393-e0a9-e50e24dcca9e" }
          ],
          "characteristics": [
            {
              "characteristic": "2a00",
              "properties": {
                "Write": "Write",
                "Read": "Read"
              },
              "service": "1800"
            },
            {
              "characteristic": "2a01",
              "properties": {
                "Read": "Read"
              },
              "service": "1800"
            },
            {
              "characteristic": "2a04",
              "properties": {
                "Read": "Read"
              },
              "service": "1800"
            },
            {
              "characteristic": "2aa6",
              "properties": {
                "Read": "Read"
              },
              "service": "1800"
            },
            {
              "characteristic": "6e400002-b5a3-f393-e0a9-e50e24dcca9e",
              "properties": {
                "Write": "Write",
                "WriteWithoutResponse": "WriteWithoutResponse"
              },
              "service": "6e400001-b5a3-f393-e0a9-e50e24dcca9e"
            },
            {
              "descriptors": [
                {
                  "value": null,
                  "uuid": "2902"
                }
              ],
              "characteristic": "6e400003-b5a3-f393-e0a9-e50e24dcca9e",
              "properties": {
                "Notify": "Notify"
              },
              "service": "6e400001-b5a3-f393-e0a9-e50e24dcca9e"
            }
          ]
        }
        ```
    6. Call `BleManager.startNotification()` (or `BleManager.startNotificationUseBuffer()`)
2. Attempts to communicate with NUS Bluetooth Peripheral:
    1. Generic steps:
        1. Encode data to be written (ex: `"REQ SCREEN HEX"` => `[82,69,81,32,83,67,82,69,69,78,32,72,69,88]`)
        2. Write data to RX Characteristic (ex: `BleManager.writeWithoutResponse()` to `6e400002-b5a3-f393-e0a9-e50e24dcca9e` -- this does not throw an error, however, :grey_question: I cannot verify the device received anything :grey_question:)
        3. Handle response (ex: listener for `BleManagerDidUpdateValueForCharacteristic` should trigger with response data, but :x: **`BleManagerDidUpdateValueForCharacteristic` never triggers** :x:)
    2. Thinking that maybe I need to write the data as something resembling UART data frames, I retried the generic steps of `2.1` with the following variations of the encoding step, `2.1.1`. None of them resulted in a notification being received:
        1. Try adding a stop byte (LF) to the end of data (ex: `[82,69,81,32,83,67,82,69,69,78,32,72,69,88,10]`) :x:
        2. Try adding start bytes (`0xff`,`0x55`) with stop byte (ex: `[255,85,82,69,81,32,83,67,82,69,69,78,32,72,69,88,10]`) :x:
        3. Try adding start bytes with data length header and no end byte (ex: `[255,85,14,82,69,81,32,83,67,82,69,69,78,32,72,69,88]`) :x:
        4. Try adding start bytes and end byte with data length header (ex: `[255,85,14,82,69,81,32,83,67,82,69,69,78,32,72,69,88,10]`) :x:
        5. Try start and end bytes, only (ex: `[255,82,69,81,32,83,67,82,69,69,78,32,72,69,88,10]`) :x:
    3. Attempt to repeat all combinations of steps `2.1` and `2.2`, but use `write()`, followed by `read()`, instead of `writeWithoutResponse()`
        1. :x: All of these fail on `read()` with an error: `Read failed` :x:
    4. Attempt `write()` followed by `readDescriptor()` on the undocumented descriptor with id=`2902`
        - This does not cause an error, but the response appears to be useless/empty 

**Expected behavior**
I would expect to receive a notification sometime after writing a command, or I would expect a write+read to work as generally expected.

**Things that work (or appear to work)**
- [x] Scan, connect, and request services from the peripheral
- [x] `BleManager.startNotification()` and `BleManager.startNotificationUseBuffer()` return without error
- [x] `BleManager.write()` and `BleManager.writeWithoutResponse()` to all advertised RX characteristics (no errors)
- [x] `BleManager.read()` from other advertised-but-not-documented characteristics returns data, but the data is static and contains things like the product name.
  - This leads me to believe that either:
    1. `react-native-ble-manager` is be working fine and I need to handle implementing the NUS protocol on top of ble-manager to buffer/packetize `write()`s and buffer/parse `read()`s/notifications?
    9. Maybe `react-native-ble-manager`

**Smartphone (please complete the following information):**
 - Device: [Fire HD 8 (10th generation)](https://developer.amazon.com/docs/fire-tablets/ft-device-specifications-firehd-models.html?v=firehd8_2020)
 - OS: Android 9, API Level 28 ([Fire OS 7.3.2.7](https://developer.amazon.com/docs/fire-tablets/ft-identifying-tablet-devices.html))
 - react-native-ble-manager version: `10.1.0`
 - react-native version: `0.71.8`

**Additional context**
- Hardware notes:
  - 2 peripherals have been used for testing: one used, one new, and both work with existing apps in the app stores (they are "smart" stopwatches)
- I am familiar with UART, but I am not entirely sure what NUS does outside of splitting payloads into sequential 20-byte `write()` calls (and the same for `read()`s), nor do I know how much of NUS I need to understand to be able to communicate with a NUS-based device using ble-manager.
- Potentially-related issues(?):
  -  Similar issues reading from device, but no details provided by the author: #1033
  - This could be similar, if my issue is related to the `write()` and/or `writeWithoutResponse()` calls (ie: not related to the notification listeners or `read()` call), but I am not sure, and there is no answer, yet: #1025
  - Worthy of note, but not directly related to my issue: I have noticed that the calls to `disconnect()` on android are not successful (no errors, but I cannot reconnect to the peripheral using another device until power-cycling the peripheral): #985

**Comment by cahna** [comment]
Accidentally submitted before I completed writing the report... standby while I edit the original post. Thanks.

**Comment by cahna** [comment]
Closing this. As -- sort of -- expected, my issue was not related to the library.

For reference: I found [a line of code in `Android-nRF-Toolbox`](https://github.com/NordicSemiconductor/Android-nRF-Toolbox/blob/05f6d9e680ca03968347d046f6d306bb015d31e9/profile_uart/src/main/java/no/nordicsemi/android/uart/data/UARTManager.kt#L110) where the MTU is explicitly set to `517`. After adding `await BleManager.requestMTU(pid, 517)`, everything started working.

**Comment by mrrogge** [comment]
Thank you so, so much for the extensive documentation on this issue! I've been battling a similar problem on an nRF5340 for a while now. In my case, the `requestMTU()` call would hang indefinitely, but once I added a 1sec delay after `connect()`, I started getting notifications.

I wonder if it would be worthwhile building an artificial delay right into the `connect()` method, to help alleviate these timing issues for others in the future. If I get a chance, I'll work on a PR.
